### DISCUSSION TITLE
Difference between 'conf-thresh' and 'iou-thresh'

### DISCUSSION THREAD

**Discussion by aonsafdar** [ultralytics/yolov5#7906] [Q&A] [answered]
Hello,
I am training yolov5s with a custom dataset. I have trouble understanding the difference between 'conf-thres' and 'iou_thres' in val.py and detect.py. My question(s):-
1) What is the difference between the two?
2) Why the default value is different in val.py and detect.py?
3) What are the general effects of altering these defaults?
Thanks.

In **val.py** 
parser.add_argument('**--conf-thres'**, type=float, **default=0.25**, help='confidence threshold')
parser.add_argument(**'--iou-thres'**, type=float, **default=0.45**, help='NMS IoU threshold')

In **detect.py**
**conf_thres=0.001**,  # confidence threshold
**iou_thres=0.6**,  # NMS IoU threshold

**Comment by glenn-jocher** [comment, marked as answer]
@aonsafdar 👋 Hello, thanks for asking about the differences between [train.py](https://github.com/ultralytics/yolov5/blob/master/train.py), [detect.py](https://github.com/ultralytics/yolov5/blob/master/detect.py) and [val.py](https://github.com/ultralytics/yolov5/blob/master/val.py) in YOLOv5 🚀.

These 3 files are designed for different purposes and utilize different dataloaders with different settings. train.py dataloaders are designed for a speed-accuracy compromise, val.py is designed to obtain the best mAP on a validation dataset, and detect.py is designed for best real-world inference results. A few important aspects of each:


## train.py

- **trainloader** LoadImagesAndLabels(): designed to load train dataset images and labels. Augmentation capable and enabled. https://github.com/ultralytics/yolov5/blob/7ee5aed0b3fa6a805afb0a820c40bbcbf29960de/train.py#L210-L213 **val_loader** LoadImagesAndLabels(): designed to load val dataset images and labels. Augmentation capable but disabled. https://github.com/ultralytics/yolov5/blob/7ee5aed0b3fa6a805afb0a820c40bbcbf29960de/train.py#L220-L223
- image size: `640`
- rectangular inference: `False`
- [confidence threshold](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L291): `0.001`
- [iou threshold](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L292): `0.6`
- [multi-label](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L120): `True`
- [padding](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L83-L90): `None`

## val.py

- **dataloader** LoadImagesAndLabels(): designed to load train, val, test dataset images and labels. Augmentation capable but disabled. 
- https://github.com/ultralytics/yolov5/blob/7ee5aed0b3fa6a805afb0a820c40bbcbf29960de/val.py#L152-L153
- image size: `640`
- rectangular inference: `True`
- [confidence threshold](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L291): `0.001`
- [iou threshold](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L292): `0.6`
- [multi-label](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L120): `True`
- [padding](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L83-L90): `0.5 * maximum stride`


## detect.py

- **dataloaders** (multiple): designed for loading multiple types of media (images, videos, globs, directories, streams). https://github.com/ultralytics/yolov5/blob/7ee5aed0b3fa6a805afb0a820c40bbcbf29960de/detect.py#L120-L128
- image size: `640`
- rectangular inference: `True`
- [confidence threshold](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/detect.py#L152): `0.25` 
- [iou threshold](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/detect.py#L153): `0.45`
- [multi-label](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/utils/general.py#L415): `False`
- [padding](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/detect.py#L45-L54): `None`


## YOLOv5 PyTorch Hub Inference

YOLOv5 PyTorch Hub models are`AutoShape()` instances used for image loading, preprocessing, inference and NMS. For more info see [YOLOv5 PyTorch Hub Tutorial](https://github.com/ultralytics/yolov5/issues/36)
https://github.com/ultralytics/yolov5/blob/7ee5aed0b3fa6a805afb0a820c40bbcbf29960de/models/common.py#L276-L282

- image size: `640`
- rectangular inference: `True`
- [confidence threshold](https://github.com/ultralytics/yolov5/blob/abfcf9eb79877971acd238cafe6149711c5056ad/models/common.py#L226-L230): `0.25 `
- [iou threshold](https://github.com/ultralytics/yolov5/blob/abfcf9eb79877971acd238cafe6149711c5056ad/models/common.py#L226-L230): `0.45`
- [multi-label](https://github.com/ultralytics/yolov5/blob/abfcf9eb79877971acd238cafe6149711c5056ad/models/common.py#L288): `False`
- [padding](https://github.com/ultralytics/yolov5/blob/abfcf9eb79877971acd238cafe6149711c5056ad/models/common.py#L226-L230): `None`

Good luck 🍀 and let us know if you have any other questions!

**Reply by aonsafdar** [reply]
Thank you for the detailed reply. 
Can you please explain the difference between the confidence threshold and Iou threshold? I apologize if it is a very basic question.

**Reply by zxcamazing** [reply]
Google search

**Reply by aonsafdar** [reply]
"confidence threshold is the minimum score that the model will consider the prediction to be a true prediction (otherwise it will ignore this prediction entirely). IoU threshold is the minimum overlap between ground truth and prediction boxes for the prediction to be considered a true positive." 
-courtesy
https://stackoverflow.com/questions/68527004/selecting-an-iou-and-confidence-threshold-for-evaluation-of-model-performance#:~:text=Just%20to%20clarify%2C%20confidence%20threshold,be%20considered%20a%20true%20positive.

**Reply by Alberto1404** [reply]
Regarding this... How should be define conf-thres and iou-thres for inference (when running detect.py) ? I mean, I suppose that based on a custom training, the graphs would tell us which value optimize it, but, i don't know how to understand them. Here I show my training results for person detection (only one class). 
Thank you in advance. Any help is appreciated
![confusion_matrix](https://user-images.githubusercontent.com/91409881/203307399-dc99dd02-2c2b-4fef-bd61-a930b2b7b17b.png)
![F1_curve](https://user-images.githubusercontent.com/91409881/203307418-94b681f3-e496-41f2-a653-20e454d99fe3.png)
![P_curve](https://user-images.githubusercontent.com/91409881/203307426-05f9abb3-7a45-4393-b667-aba88a44eaef.png)
![PR_curve](https://user-images.githubusercontent.com/91409881/203307436-54ccee6f-f92d-45bf-9a8e-0a4378f96e76.png)
![R_curve](https://user-images.githubusercontent.com/91409881/203307445-74dac9dd-4a4c-436b-9d62-c09438e83f94.png)
![results](https://user-images.githubusercontent.com/91409881/203307458-24c24f63-e927-4591-af4f-aed6306185b9.png)

**Reply by glenn-jocher** [reply]
@Alberto1404 you can use max F1 curve to determine your best confidence threshold. For IoU you can use the default or try different settings in val.py to see it's effect on mAP.

**Reply by yesid-acm** [reply]
@aonsafdar  @glenn-jocher why does a warning to use confi >0.001 appear in val.py?

to better use yolov5 you should know a little more in the tutorial, add the step by step. For example: train with XX thresholds and use those in the val.py which are the best, etc... don't you think glenn?

**Reply by lucasvw** [reply]
> you can use max F1 curve to determine your best confidence threshold.

@glenn-jocher what's your intuition behind this? I can confirm on my custom trained yolo model that this indeed gives best results on the test set, but I'm failing to see the intuition behind this. I have tried searching for this online, but can't seem to find a good explanation. Perhaps you can share your ideas or perhaps you can point me to a resource?

**Reply by faisalhazry** [reply]
@lucasvw 
**Precision:** Knowns as ability of a model to correctly identify positive predictions out of instances predicted as positive **Precision = TP / (TP + FP)**

**Recall:** Known as sensitivity of true positive rate by total total relevant results  instances correctly classified  **Recall = TP / (TP + FN)**

In conclusion we cannot possible to maximize both metric recall and precision at the same time as one comes at the cost of another. To analysis precision and recall we need to balancing two separate scores

F-1 scores which is harmonic precision and recall camas help to give a good simplicity matrix for evaluation model.

**F-1 scores**: Increasing the F1 score indicates an improvement in the overall performance of a model

Confidence -> is  IoU Overlapping result. This one not sure 

Reference `https://medium.com/@shrutisaxena0617/precision-vs-recall-386cf9f89488` 

@glenn-jocher plz correct  me if i am wrong
